microsoft/LightGBM
# [R-package] prevent symbol lookup conflicts (fixes #4045)

This PR fixes a critical issue affecting Windows users of the R package. In #4045, several users reported a problem like "if I load certain R libraries before `{lightgbm}`, `{lightgbm}` can cause the R session to crash when constructing a dataset".

After some investigation (detailed in https://github.com/microsoft/LightGBM/issues/4045#issuecomment-812289182), I believe I found the root cause for this. The routine `R_registerRoutines()` at https://github.com/microsoft/LightGBM/blob/d517ba12f2e7862ac533908304dddbd770655d2b/R-package/src/lightgbm_R.cpp#L724 isn't actually being called by R!

The key piece of documentation is at https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Converting-a-package-to-use-registration.

> Construct the registration table ... if included in a C++ file the ‘R_init’ function would need to be declared extern "C"):

This PR adds that declaration to `lightgbm_R.cpp` and adds a unit test to prevent this type of mistake from being re-introduced in the future.

Fixes #4045.
Addresses the Windows half of #4034.

I need to test if this would fix #4007. Will update this description once I try those.

<hr>

## Background

R provides an interface for "native routine registration", a mechanism to tell R exactly which symbols it should expect to find in a shared library (`.dll` / `.so`). This makes calls to compiled code with `.Call()` safer because R looks in a specific place.

If that type of registration isn't done, then it can create the possibility for conflicts with other loaded DLLs. From https://cran.r-project.org/doc/manuals/r-release/R-exts.html#dyn_002eload-and-dyn_002eunload:

> If a shared object/DLL is loaded more than once the most recent version is used. More generally, if the same symbol name appears in several shared objects, the most recently loaded occurrence is used. The PACKAGE argument and registration (see the next section) provide good ways to avoid any ambiguity in which occurrence is meant.

This note from https://cran.r-project.org/doc/manuals/r-release/R-exts.html#Registering-native-routines is relevant as well.

> In calls to .C, .Call, .Fortran and .External, R must locate the specified native routine by looking in the appropriate shared object/DLL. By default, R uses the operating-system-specific dynamic loader to lookup the symbol in all132 loaded DLLs and the R executable or libraries it is linked to. Alternatively, the author of the DLL can explicitly register routines with R and use a single, platform-independent mechanism for finding the routines in the DLL. One can use this registration mechanism to provide additional information about a routine, including the number and type of the arguments, and also make it available to R programmers under a different name.

<hr>

## How to test this

You can run the following code to check if registration was successful.

```r
library(lightgbm)
dll_info <- getLoadedDLLs()[["lightgbm"]]

# check that dynamic lookup has been disabled
!isTRUE(dll_info[["dynamicLookup"]])

# check that all the expected symbols show up
getDLLRegisteredRoutines(dll_info[["path"]])
```

```text
                               .Call .Call.numParameters
1                LGBM_GetLastError_R                   3
2       LGBM_DatasetCreateFromFile_R                   5
3        LGBM_DatasetCreateFromCSC_R                  10
4        LGBM_DatasetCreateFromMat_R                   7
5            LGBM_DatasetGetSubset_R                   6
6      LGBM_DatasetSetFeatureNames_R                   3
7      LGBM_DatasetGetFeatureNames_R                   5
8           LGBM_DatasetSaveBinary_R                   3
9                 LGBM_DatasetFree_R                   2
10            LGBM_DatasetSetField_R                   5
11        LGBM_DatasetGetFieldSize_R                   4
12            LGBM_DatasetGetField_R                   4
13 LGBM_DatasetUpdateParamChecking_R                   3
14          LGBM_DatasetGetNumData_R                   3
15       LGBM_DatasetGetNumFeature_R                   3
16              LGBM_BoosterCreate_R                   4
17                LGBM_BoosterFree_R                   2
18 LGBM_BoosterCreateFromModelfile_R                   3
19 LGBM_BoosterLoadModelFromString_R                   3
20               LGBM_BoosterMerge_R                   3
21        LGBM_BoosterAddValidData_R                   3
22   LGBM_BoosterResetTrainingData_R                   3
23      LGBM_BoosterResetParameter_R                   3
24       LGBM_BoosterGetNumClasses_R                   3
25       LGBM_BoosterUpdateOneIter_R                   2
26 LGBM_BoosterUpdateOneIterCustom_R                   5
27     LGBM_BoosterRollbackOneIter_R                   2
28 LGBM_BoosterGetCurrentIteration_R                   3
29  LGBM_BoosterGetUpperBoundValue_R                   3
30  LGBM_BoosterGetLowerBoundValue_R                   3
31        LGBM_BoosterGetEvalNames_R                   5
32             LGBM_BoosterGetEval_R                   4
33       LGBM_BoosterGetNumPredict_R                   4
34          LGBM_BoosterGetPredict_R                   4
35      LGBM_BoosterPredictForFile_R                  11
36      LGBM_BoosterCalcNumPredict_R                   9
37       LGBM_BoosterPredictForCSC_R                  15
38       LGBM_BoosterPredictForMat_R                  12
39           LGBM_BoosterSaveModel_R                   5
40   LGBM_BoosterSaveModelToString_R                   7
41           LGBM_BoosterDumpModel_R                   
```
